### PR TITLE
Ported `go_grader`&`grades` into faasten

### PR DESCRIPTION
**Pull request by tan-yue**
This PR contains ported `go_grader`&`grades` and changes to test them.
_TLDR:_
`./test_ported_functions.sh` should build the correct versions of binaries & filesystem images, test run the two functions, and `cat` the two functions' output files to the `stdout`.

Changes are:
* Porting includes replacing `write_key` and `read_key` with `fsread`, `fscreate_[dir|file]` and `fswrite` and calling `endorse_with` and `declassify_to`.
* The two example payloads also are modified to include `user` and `function` keys in the `context` so that a function can use the values to label files it creates and to exercise its privilege.
*  `faasten` branch of `snapfaas` includes a new binary `sffs` that supports interactions with the labeled file system from the command line.
* Makefile now includes a `prepfs` target. The target recipe creates directories and files which test runs of functions assume exist using `sffs`. The recipe also tells how these files and directories are labeled.
* test_ported_functions.sh should be self-explanatory. It pulls the faasten version `snapfaas` and `snapfaas-images` repositories and builds the binaries and filesystem. Then it makes use of Makefile to test run the ported functions. Last, it `cat`s to the `stdout` the files each function writes to the labeled file system.

### CODE REVIEW

File: functions/gh_repo/workload.py [hunk 1]
```diff
@@ -4,17 +4,23 @@
...
-    metadataString = syscall.read_key(bytes(meta_key, "utf-8")) or "{}"
+    metadataString = syscall.fsread(meta_path) or "{}"
     if metadataString:
         metadata = json.loads(metadataString)
-        workflow = json.loads(syscall.read_key(bytes(workflow_key, "utf-8")) or "[]")
+        workflow = json.loads(syscall.fsread(workflow_path) or "[]")
 
         resp = syscall.github_rest_get("/repos/%s/tarball/%s" % (req["repository"]["full_name"], req["after"]));
-        syscall.write_key(bytes(key, "utf-8"), resp.data)
+        syscall.fscreate_file(base_dir, '%s.tgz' % req['after'], syscall.get_current_label())
```
Comment (alevy) [inline functions/gh_repo/workload.py]: Ideally, here we would label the output file with secrecy corresponding to the owners of the repository.

File: functions/grades/workload.py [hunk 2]
```diff
@@ -45,10 +45,17 @@ def app_handle(args, context, syscall):
...
     }
 
-    key = os.path.join(os.path.dirname(args["test_results"]),"grade.json")
-    syscall.write_key(bytes(key, "utf-8"), bytes(json.dumps(output), "utf-8"))
+    user = context['user']
+    func = context['function']
+    base_dir = os.path.join('/', func, user, context['metadata']['assignment'])
+    file = os.path.join(base_dir, 'grade.json')
+    syscall.endorse_with([[func]])
+    target_label = syscall.new_dclabel([[user]], [[func]])
```
Comment (alevy) [inline functions/grades/workload.py]: How does this `target_label` relate to the `current_label`? Ideally the current label in this function would more or have _at least_ the right secrecy component at this point (whatever data it reads should be exactly secret to the same people who can read the final grades)

**Comment by alevy** [comment]
A description (potentially as a comment at the top of each function or something, plus some overall description somewhere) of what labels each function expects and how it modifies the current label (through reads or explicit tainting/endorsing) would be really useful.

**Comment by tan-yue** [comment]
@alevy 
I pushed two new commits, [the first one is the major one to look at](https://github.com/alevy/sf-grader/pull/1/commits/1a143776e60ace2fd06dea3ea843df8ef9a68c85).

In this commit, I switched to a new set of file system syscalls `workspace_xxx` and changed `declassify`'s semantics*.

1. The idea of workspace_xxx syscalls is that we expect functions to if not only but most of time in their workspace directories, `/$function/$user1_$user2_..._$userN`. Particularly, functions can only write their workspace directories. Functions can read paths in other functions'  workspaces with the restriction that such paths has to belong to the current end users.
2. Another big change is how endorsement is handled. I removed endorse syscall and have write syscalls `workspace_[createdir|createfile|write]` always endorse with the function's own principal**.
3. Create syscalls `workspace_[createdir|createfile]` now have its label argument default to None. That is, by default, creation labels the newly created file with the function's current label**.
4. `declassify` now takes in a DcComponent type argument, secrecy. Since there's no syscall for the function to construct a DcComponent from scratch, secrecy should be a secrecy remembered by the function of a label returned by the syscall `get_current_label`.
5. I renamed original file system write and read syscalls to `fs_read|write|createdir|createfile`. These syscalls use absolute path starting with the root '/'. Ported `go_grader` and `grades` both only uses `fs_read` over paths passed in from their payloads. I am not sure how the rest write syscalls can be useful. I simply keep them there and we can remove them in the future.

*See Python's [syscalls.py](https://github.com/princeton-sns/snapfaas-images/blob/84f9e85e7f90a7ac9e9cf6079bbc32cf9aa3293a/rootfs/faasten/runtimes/python3/syscalls.py).
**See syscall handlers of FsCreate, FsRead, FsWrite in VMM, [vm.rs](https://github.com/princeton-sns/snapfaas/blob/7cc0dd706fbfce07f00f2c8dd72609941931e5ae/snapfaas/src/vm.rs)

**Comment by tan-yue** [comment]
@alevy 
The latest commit depends on these two PRs: [snapfaas](https://github.com/princeton-sns/snapfaas/pull/32) and [snapfaas-images](https://github.com/princeton-sns/snapfaas-images/pull/19).

`go_grader` and `grades` now use `open|create_unnamed` to pass down intermediate data to the direct downstream functions.
